File: non_native/lib/views/edit_flashcard_screen.dart

import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../view_models/main_view_model.dart';
import '../models/flashcard.dart';
import '../assets/difficulty.dart';
import 'difficulty_dropdown.dart';

class EditFlashcardScreen extends StatefulWidget {
  final int flashcardId;
  const EditFlashcardScreen({super.key, required this.flashcardId});

  @override
  State<EditFlashcardScreen> createState() => _EditFlashcardScreenState();
}

class _EditFlashcardScreenState extends State<EditFlashcardScreen> {
  late Flashcard _initialFlashcard;
  late TextEditingController _questionController;
  late TextEditingController _answerController;
  late Difficulty _selectedDifficulty;
  String? _errorQuestion;
  String? _errorAnswer;

  @override
  void initState() {
    super.initState();
    
    final viewModel = context.read<MainViewModel>();
    _initialFlashcard = viewModel.getFlashcardById(widget.flashcardId)!;

    
    _questionController = TextEditingController(text: _initialFlashcard.question);
    _answerController = TextEditingController(text: _initialFlashcard.answer);
    _selectedDifficulty = _initialFlashcard.difficulty;
  }

  void _validateQuestion(String input) {
    if (input.trim().isEmpty) {
      _errorQuestion = "Question cannot be empty";
    } else if (input.length > 150) {
      _errorQuestion = "Maximum 150 characters";
    } else {
      _errorQuestion = null;
    }
  }

  void _validateAnswer(String input) {
    if (input.trim().isEmpty) {
      _errorAnswer = "Answer cannot be empty";
    } else if (input.length > 150) {
      _errorAnswer = "Maximum 150 characters";
    } else {
      _errorAnswer = null;
    }
  }

  void _validateAndSave(MainViewModel viewModel) {
    
    setState(() {
      _validateQuestion(_questionController.text);
      _validateAnswer(_answerController.text);
    });

    if (_errorQuestion == null && _errorAnswer == null) {
      
      final updatedFlashcard = _initialFlashcard.copyWith(
        question: _questionController.text.trim(),
        answer: _answerController.text.trim(),
        difficulty: _selectedDifficulty,
      );

      
      viewModel.updateFlashcard(updatedFlashcard);
      Navigator.of(context).pop(); 
    }
  }

  @override
  Widget build(BuildContext context) {
    final viewModel = context.read<MainViewModel>();

    return Scaffold(
      backgroundColor: const Color(0xFFF0F0F0),
      appBar: AppBar(
        title: const Text("Edit card"),
        backgroundColor: const Color(0xFFF0F0F0),
        centerTitle: true,
        leading: IconButton(
          icon: const Icon(Icons.close),
          onPressed: () => Navigator.of(context).pop(), 
        ),
        bottom: PreferredSize(
          preferredSize: const Size.fromHeight(1.0),
          child: Divider(color: Colors.black.withOpacity(0.07), thickness: 1.0),
        ),
      ),
      body: SingleChildScrollView(
        padding: const EdgeInsets.all(16.0),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            
            const Text("Question", style: TextStyle(fontWeight: FontWeight.bold, fontSize: 13.0)),
            const SizedBox(height: 8.0),
            TextFormField(
              controller: _questionController,
              onChanged: (value) {
                
                if (_errorQuestion != null) setState(() => _errorQuestion = null);
              },
              maxLines: null,
              minLines: 5,
              maxLength: 150,
              textCapitalization: TextCapitalization.sentences,
              style: const TextStyle(fontSize: 18.0),
              decoration: InputDecoration(
                hintText: "e.g. What is the capital of Japan?",
                border: const OutlineInputBorder(),
                errorText: _errorQuestion,
                focusedBorder: const OutlineInputBorder(borderSide: BorderSide(color: Color(0xFF1280ED))),
                counterText: '',
              ),
            ),
            const SizedBox(height: 12.0),

            
            const Text("Answer", style: TextStyle(fontWeight: FontWeight.bold, fontSize: 13.0)),
            const SizedBox(height: 8.0),
            TextFormField(
              controller: _answerController,
              onChanged: (value) {
                
                if (_errorAnswer != null) setState(() => _errorAnswer = null);
              },
              maxLines: null,
              minLines: 4,
              maxLength: 150,
              textCapitalization: TextCapitalization.sentences,
              style: const TextStyle(fontSize: 18.0),
              decoration: InputDecoration(
                hintText: "e.g. Tokyo",
                border: const OutlineInputBorder(),
                errorText: _errorAnswer,
                focusedBorder: const OutlineInputBorder(borderSide: BorderSide(color: Color(0xFF1280ED))),
                counterText: '',
              ),
            ),
            const SizedBox(height: 12.0),

            
            DifficultyDropdown(
              selectedDifficulty: _selectedDifficulty,
              onDifficultySelected: (newDifficulty) {
                setState(() {
                  _selectedDifficulty = newDifficulty;
                });
              },
            ),
          ],
        ),
      ),
      bottomNavigationBar: Padding(
        padding: EdgeInsets.only(
          left: 25.0,
          right: 25.0,
          top: 25.0,
          bottom: 25.0 + MediaQuery.of(context).viewInsets.bottom,
        ),
        child: Row(
          children: [
            Expanded(
              child: ElevatedButton(
                onPressed: () => Navigator.of(context).pop(), 
                style: ElevatedButton.styleFrom(
                  backgroundColor: Colors.grey[300],
                  padding: const EdgeInsets.symmetric(vertical: 15.0),
                ),
                child: const Text("Cancel", style: TextStyle(color: Colors.black)),
              ),
            ),
            const SizedBox(width: 16.0),
            Expanded(
              child: ElevatedButton(
                onPressed: () => _validateAndSave(viewModel), 
                style: ElevatedButton.styleFrom(
                  backgroundColor: const Color(0xFF1E88E5),
                  padding: const EdgeInsets.symmetric(vertical: 15.0),
                ),
                child: const Text("Save", style: TextStyle(color: Colors.white)),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

File: Native/src/main/java/com/example/mindflip_native/views/EditFlashcardScreen.kt
package com.example.mindflip_native.views

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.mindflip_native.assets.Difficulty
import com.example.mindflip_native.models.Flashcard
import com.example.mindflip_native.viewModels.MainViewModel

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun EditFlashcardScreen(
    viewModel: MainViewModel,
    flashcardId: Int,
    onCancel: () -> Unit,
    onSave: () -> Unit
) {
    val flashcard = viewModel.flashcards.value?.find { it.localId == flashcardId }
    if (flashcard == null) {
        Text("Flashcard not found")
        return
    }

    var question by remember { mutableStateOf(flashcard.question) }
    var answer by remember { mutableStateOf(flashcard.answer) }
    var difficulty by remember { mutableStateOf(flashcard.difficulty) }
    var errorQuestion by remember { mutableStateOf("") }
    var errorAnswer by remember { mutableStateOf("") }

    fun validateQuestion(input: String) {
        errorQuestion = when {
            input.isBlank() -> "Question cannot be empty"
            input.length > 150 -> "Maximum 150 characters"
            else -> ""
        }
    }

    fun validateAnswer(input: String) {
        errorAnswer = when {
            input.isBlank() -> "Answer cannot be empty"
            input.length > 150 -> "Maximum 150 characters"
            else -> ""
        }
    }

    Scaffold(
        topBar = {
            Column {
                CenterAlignedTopAppBar(
                    title = { Text("Edit card") },
                    navigationIcon = {
                        IconButton(onClick = onCancel) {
                            Icon(Icons.Default.Close, contentDescription = "Back")
                        }
                    },
                    colors = TopAppBarDefaults.centerAlignedTopAppBarColors(
                        containerColor = Color(0xFFF0F0F0)
                    )
                )
                Divider(
                    color = Color.Black.copy(alpha = 0.07f),
                    thickness = 1.dp
                )
            }
        },
        bottomBar = {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(25.dp)
                    .imePadding(),
                horizontalArrangement = Arrangement.spacedBy(16.dp)
            ) {
                Button(
                    onClick = onCancel,
                    colors = ButtonDefaults.buttonColors(
                        containerColor = Color.LightGray
                    ),
                    modifier = Modifier.weight(1f)
                ) {
                    Text("Cancel", color = Color.Black)
                }

                Button(
                    onClick = {
                        validateQuestion(question)
                        validateAnswer(answer)

                        if (errorQuestion.isEmpty() && errorAnswer.isEmpty()) {
                            viewModel.updateFlashcard(
                                flashcard.copy(
                                    question = question,
                                    answer = answer,
                                    difficulty = difficulty
                                )
                            )
                            onSave()
                        }
                    },
                    colors = ButtonDefaults.buttonColors(
                        containerColor = Color(0xFF1E88E5)
                    ),
                    modifier = Modifier.weight(1f)
                ) {
                    Text("Save", color = Color.White)
                }
            }
        },
        containerColor = Color(0xFFF0F0F0)
    ) { paddingValues ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues)
                .padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp)
        ) {
            // Question
            Text("Question", style = MaterialTheme.typography.labelMedium)
            OutlinedTextField(
                value = question,
                onValueChange = {
                    question = it
                    validateQuestion(it)
                },
                placeholder = {
                    Text(
                        "e.g. What is the capital of Japan?",
                        color = Color.Black.copy(alpha = 0.3f)
                    )
                },
                isError = errorQuestion.isNotEmpty(),
                singleLine = false,
                textStyle = LocalTextStyle.current.copy(
                    fontSize = 18.sp
                ),
                keyboardOptions = KeyboardOptions.Default.copy(
                    capitalization = KeyboardCapitalization.Sentences
                ),
                modifier = Modifier
                    .fillMaxWidth()
                    .height(150.dp),
                colors = OutlinedTextFieldDefaults.colors(
                    focusedBorderColor = Color(0xFF1280ED)
                )
            )
            if (errorQuestion.isNotEmpty()) {
                Text(errorQuestion, color = Color.Red, style = MaterialTheme.typography.bodySmall)
            }

            // Answer
            Text("Answer", style = MaterialTheme.typography.labelMedium)
            OutlinedTextField(
                value = answer,
                onValueChange = {
                    answer = it
                    validateAnswer(it)
                },
                placeholder = {
                    Text("e.g. Tokyo", color = Color.Black.copy(alpha = 0.3f))
                },
                isError = errorAnswer.isNotEmpty(),
                singleLine = false,
                textStyle = LocalTextStyle.current.copy(
                    fontSize = 18.sp
                ),
                keyboardOptions = KeyboardOptions.Default.copy(
                    capitalization = KeyboardCapitalization.Sentences
                ),
                modifier = Modifier
                    .fillMaxWidth()
                    .height(110.dp),
                colors = OutlinedTextFieldDefaults.colors(
                    focusedBorderColor = Color(0xFF1280ED)
                )
            )
            if (errorAnswer.isNotEmpty()) {
                Text(errorAnswer, color = Color.Red, style = MaterialTheme.typography.bodySmall)
            }

            // Difficulty Dropdown
            DifficultyDropdown(
                selectedDifficulty = difficulty,
                onDifficultySelected = { difficulty = it }
            )
        }
    }
}
